src/Features/CSharp/Portable/ConvertAutoPropertyToFullProperty/CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -30,26 +31,14 @@ public CSharpConvertAutoPropertyToFullPropertyCodeRefactoringProvider()
         {
         }
 
-        internal override SyntaxNode GetProperty(SyntaxToken token)
+        internal override async Task<SyntaxNode> GetPropertyAsync(Document document, TextSpan span, CancellationToken cancellationToken)
         {
-            var containingProperty = token.Parent.FirstAncestorOrSelf<PropertyDeclarationSyntax>();
-            if (containingProperty == null)
-            {
-                return null;
-            }
 
-            if (!(containingProperty.Parent is TypeDeclarationSyntax))
-            {
-                return null;
-            }
+            var refactoringHelperService = document.GetLanguageService<IRefactoringHelpersService>();
 
-            var start = containingProperty.AttributeLists.Count > 0
```

Comment:
Not quite sure what the idea here was. If I understand that correctly it was enabling it when cursor was anywhere in Attributes for the Property. That said: 
1) No tests for that
2) Haven't seen that anywhere else yet

But again, breaking change & so let's ask "is it desirable?"

Reply:
The original behavior was intentional.  It said:  do *not* enable inside the attirbutes.  i.e. it sets the valid start position *after* the end of teh attributes.

Please don't break this.  Or, rather, let's think through this.  All these cases should be supported

1. the caret is in the prop-header, but not in the attribute.
2. user selects the whole prop, but not the attributes.
3. user selects the whole prop, including the attributes.

Any selection *within* the attributes should not count as the selection of the thing that owns the attributes.

Note: this logic should be uniform across all things that can have attributes.  It would be *very* desirable for it to be in the refactoring helper.

Reply:
Yep, you're right, it disables the refactoring when caret is within attributes.

1) We're currently breaking that see & discuss above (`public int[||] Goo { get; set; }`
2), 3) Let's continue discussion & testing for that here

Note: I wholeheartedly agree with the note.

Reply:
Is there some dummy / always in scope Attribute one could use in tests or should the test include just a `DummyAttribute : Attribute`?

Reply:
There're also a few new edge cases:
4) User is on the (left) edge of prop header that is preceded by attribute (current edge climbing won't climb because the declaration Node has span including the attribute)
5) User is on the edge of attribute (left), do we really want to climb & "return" the prop declaration node? 

6) How do these situation change for attributes that are inline

Reply:
I'll have to test it properly later but 3) should work fine as well as 2). In both cases the smallest encompassing node will still be the wanted one.

Reply:
Why do attributes need to be handled specially? I think we can achieve all what is discussed here with following simple logic, which doesn’t depend on special casing for any node kinds:

1. A token can be the first token of multiple nodes.
2. If the caret is just before the first token ‘t’ of the parent node, keep walking up the node parent chain until either ‘t’ is no longer the first token of the current node OR current node is a candidate that should be returned.

The above should analogously be applied for caret just after last tokens too.

Reply:
Looking at your example 

```csharp
    [||][Dummy]
    [||]int A{ get; set; }

```

we do need to handle the case where caret is just before a token, which is the first token of a node ignoring attributes and doc comments (i.e. first token of effective node header)

Reply:
Yep, we might also want to explicitly not allow (TBH I have no preference / intuition on this one):
```
[||][Dummy]
    int A{ get; set; }
```

The other cases should be handle by design (need to test them to make sure).

Reply:
> we might also want to explicitly not allow 

I think we need to allow when caret is before the attribute list, otherwise the following will not show any property refactorings:
`[||][Dummy] int A{ get; set; }`

Reply:
Speaking of which, is there a way to "ignore/extract" attributes (docs comments are trivia, we already ignore that)? I've looked into SyntaxFacts and there're little to no general Attribute helpers.

I thought about going through parent's children as long as the Node is `AttributeList` but am actually not sure if the Children are guaranteed to be properly ordered.

Reply:
I don't know of any such existing helper, but @CyrusNajmabadi might know for sure.

Reply:
> User is on the edge of attribute (left), do we really want to climb & "return" the prop declaration node?

Yes.

Reply:
> Speaking of which, is there a way to "ignore/extract" attributes (docs comments are trivia, we already ignore that)? I've looked into SyntaxFacts and there're little to no general Attribute helpers.

Possibly.  SyntaxGenerator has helpers around attributes which might be useful to you.

> but am actually not sure if the Children are guaranteed to be properly ordered.

i'm not sure what this means. can you clarify?

Reply:
If you iterate over node's children, is the order of iteration guaranteed to be in the order of their spans? I.e. for DeclaradionNode above the children are __guaranteed__ to be iterated in following order: `AttributeList, PUblicKeyword, PredefinedType, …`

Reply:
Yes.  this is a core invariant of the roslyn syntactic model.  children are iterated in order, and children *always* (literally zero exceptions) will be in span order.  

Note: in your specific case, it would be `attribute list, modifiers, return type, name, accessor list`.

Reply:
There're a few other edge-cases that need to be taken care of:
```
    public class B : A
    {
      [|[Test]
        [Test2]|]
        void C()
        {

        }
    }
```

triggers PullMethodUp because the smallest encompassing node is the MethodDeclaration.

Reply:
One simple approach *may* be:

if the caller isn't *explicitly* asking to select the AttributeLists node, then first see if any AttributeLists are selected as per our normal algorithm.  If so, then this is *not* the selection of any other type of node.

Reply:
Thinking about it more I'm not sure I fully understand the idea. If only one AttributeList is selected then everything works as it should. And if multiple are selected then it's parent node gets returned. But then no other node should be returned since we established we don't want to try to return their immediate parent node but there's also no other node that would be the smallest encompassing one.

Reply:
> Thinking about it more I'm not sure I fully understand the idea.

The idea is simply: detect if the user is working with attributes.  If they are, then only match the request to "try get the selected thing" if the user was asking for attributes.  In other words, if we think attributes are the focus of hte user, this shouldn't count as a selection of the thing that owns the attributes.

Reply:
I believe I've handled all Attribute situations but can't be absolutely sure.